src/control/BBTrajectories/BBTrajectory1D.tpp

```diff
@@ -0,0 +1,210 @@
+//
+// Created by rolf on 15-01-20.
+//
+
+#include "control/BBTrajectories/BBTrajectory1D.h"
+#include <cmath>
+#include <iostream>
+#include <chrono>
+template<class num>
+num BBTrajectory1D<num>::fullBrakePos(num pos, num vel, num accMax) noexcept {
+    num acc = vel <= 0 ? accMax : - accMax;
+    num t = - vel/acc; // (inverted) time needed to break to zero velocity
+    return pos + 0.5*vel*t; // position after breaking
+}
+template<class num>
+num BBTrajectory1D<num>::accelerateBrakePos(num pos0, num vel0, num vel1, num accMax) noexcept {
+    num acc1;
+    num acc2;
+    if (vel1 >= vel0) {
+        acc1 = accMax;
+        acc2 = - accMax;
+    }
+    else {
+        acc1 = - accMax;
+        acc2 = accMax;
+    }
+    num t1 = (vel1 - vel0)/acc1; // time to reach vel1
+    num pos1 = pos0 + (0.5*(vel0 + vel1)*t1); //position at which we reach vel1
+    num t2 = - vel1/acc2; // time to max break from vel1 to 0
+    return pos1 + (0.5*vel1*t2); // position we stop at after initiating maximal break at pos1
+
+}
+template<class num>
+void
+BBTrajectory1D<num>::triangularProfile(num initialPos, num initialVel, num finalPos, num maxAcc,
+        bool invertedSign) noexcept {
+    num acc = invertedSign ? maxAcc : - maxAcc;
+    //compute the final time difference at which we switch from accelerating to breaking
+    num sq = (acc*(finalPos - initialPos) + 0.5*initialVel*initialVel)/(acc*acc);
+    num brakeTime = sq > 0 ? sqrt(sq) : 0;
+    num topVel = acc*brakeTime;
+    num switchTime = (topVel - initialVel)/acc;
+    num switchPos = initialPos + (initialVel + topVel)*0.5*switchTime;
+    updatePart(0, switchTime, acc, initialVel, initialPos);
+    updatePart(1, switchTime + brakeTime, - acc, topVel, switchPos);
+    numParts = 2;
+}
+template<class num>
+void BBTrajectory1D<num>::updatePart(int index, num tEnd, num acc, num vel, num pos) noexcept {
+    parts[index].acc = acc;
+    parts[index].startPos = pos;
+    parts[index].startVel = vel;
+    parts[index].tEnd = tEnd;
+}
+template<class num>
+void BBTrajectory1D<num>::trapezoidalProfile(num initialPos, num initialVel, num maxVel, num finalPos,
+        num maxAcc) noexcept {
+    num acc1;
+    num acc3;
+
+    if (initialVel > maxVel) {
+        acc1 = - maxAcc;
+    }
+    else {
+        acc1 = maxAcc;
+    }
+
+    if (maxVel > 0) {
+        acc3 = - maxAcc;
+    }
+    else {
+        acc3 = maxAcc;
+    }
+
+    num t1 = (maxVel - initialVel)/acc1;
+    num t3 = - maxVel/acc3;
+    num startCoastPos = initialPos + 0.5*(initialVel + maxVel)*t1;
```

Comment:
This doesn't make sense to me. It seems that this calculates the area of a 45,45,90 triangle, which is only valid if  `initialVel` = 0. Should it not be (given a graph where x=time, y=velocity)
`t1 * initialVel` (area of square) + `0.5 * (maxVel - initialVel) * t1` (area of triangle on top of square)`

Reply:
Be careful, we are working with integrals here so it is signed triangle area. It makes sense because over segment 1 we have average velocity (initialVel+maxVel)/2 for t1 seconds and we start at the initialPos.